src/main/java/uk/ac/cam/cl/dtg/segue/comm/MailGunEmailManager.java

```diff
@@ -0,0 +1,180 @@
+        //  - filters users by their email preferences and email verification status
+        //  - for each user U, extracts the name and email of U, and adds them to the (possibly empty) map of
+        //    `userVariables.get(U.getId())`
+        //  - creates a new map with emails as keys and the updated `userVariables` maps as values
+        Map<String, Object> recipientVariables = userDTOs.stream()
+                .filter(user -> {
+                    // FIXME - should system emails ignore this setting to avoid abuse?
+                    if (user.getEmailVerificationStatus() == EmailVerificationStatus.DELIVERY_FAILED) {
+                        return false;
+                    }
+                    try {
+                        UserPreference preference = userPreferenceManager.getUserPreference(
+                                SegueUserPreferences.EMAIL_PREFERENCE.name(), emailType.name(), user.getId());
+                        // If no preference is present, do not send email to this user.
+                        return preference != null && preference.getPreferenceValue();
+                    } catch (SegueDatabaseException e) {
+                        // Assume the worst and do not send the email to this user in the case of a DB issue
+                    }
+                    return false;
+                }).collect(
+                    Collectors.toMap(
+                        RegisteredUserDTO::getEmail,
+                        u -> Stream.of(
+                                ImmutableMap.of(
+                                "email", u.getEmail(),
+                                "givenName", u.getGivenName()
+                                ),
+                                userVariables.getOrDefault(u.getId(), ImmutableMap.of())
+                        ).flatMap(m -> m.entrySet().stream())
+                        .collect(Collectors.toMap(Entry::getKey, Entry::getValue, (e1, e2) -> e1))
+                    )
+                );
+
+        String fromAddress = emailContentTemplate.getOverrideFromAddress();
+        if (null == fromAddress || fromAddress.isEmpty()) {
+            fromAddress = globalProperties.getProperty(Constants.MAILGUN_FROM_ADDRESS);
+        }
+        String fromName = emailContentTemplate.getOverrideFromName();
+        if (null == fromName || fromName.isEmpty()) {
+            fromName = globalProperties.getProperty(Constants.MAIL_NAME);
+        }
+        String replyToAddress = emailContentTemplate.getReplyToEmailAddress();
+        if (null == replyToAddress || replyToAddress.isEmpty()) {
+            replyToAddress = globalProperties.getProperty(Constants.REPLY_TO_ADDRESS);
+        }
+        String replyToName = emailContentTemplate.getReplyToName();
+        if (replyToName == null || replyToName.isEmpty()) {
+            replyToName = globalProperties.getProperty(Constants.MAIL_NAME);
+        }
+        String replyTo = replyToName + " <" + replyToAddress + "> ";
```

Comment:
Possibly but if it works the same way I'd rather leave it like this for now

Reply:
I'm not sure they _are_ the same; email headers are complicated and email addresses are messy. But I suppose since we control the input here, it's probably okay for now.